### PR TITLE
Taklons: brainstone

### PR DESCRIPTION
**Pull request by rzulian**
Added brainstone management using heuristic

### CODE REVIEW

**Comment by coyotte508** [comment]
Related to #20

File: src/player-data.ts [hunk 1]
```diff
@@ -51,6 +51,7 @@ export default class PlayerData extends EventEmitter {
   greenFederations: number = 0;
   /** Coordinates occupied by buildings */
   occupied: GaiaHex[] = [];
+  brainstone: number = -1;
```
[the remark this answers is not in the document]
Comment (rzulian) [reply src/player-data.ts]: added enums for brainstone areas

File: src/player-data.ts [hunk 2]
```diff
@@ -147,37 +149,66 @@ export default class PlayerData extends EventEmitter {
...
     this.power.area1 -= area1ToGaia;
     this.power.area2 -= area2ToGaia;
     this.power.area3 -= area3ToGaia;
+    if (brainstoneNeeded) {
+      this.brainstone = -1;
+    }
 
     if (type === Resource.GainTokenGaiaArea) {
       this.power.gaia += area1ToGaia + area2ToGaia + area3ToGaia;
+      this.brainstone = 0;
```
[the remark this answers is not in the document]
Comment (rzulian) [reply src/player-data.ts]: corrected

File: src/player-data.ts [hunk 3]
```diff
@@ -147,37 +149,68 @@ export default class PlayerData extends EventEmitter {
...
       this.power.area3 -= power;
       this.power.area1 += power;
   }
 
   discardPower(power: number, type: Resource) {
     const area1ToGaia = Math.min(power, this.power.area1);
     const area2ToGaia = Math.min(power - area1ToGaia, this.power.area2);
-    const area3ToGaia = Math.min(power - area1ToGaia - area2ToGaia, this.power.area2);
+    const area3ToGaia = Math.min(power - area1ToGaia - area2ToGaia, this.power.area3);
+    const brainstoneNeeded = this.brainstoneInPlay() && (power - area1ToGaia - area2ToGaia - area3ToGaia >0);
```
Comment (coyotte508) [inline src/player-data.ts]: the second part can be `&& this.discardablePowerTokens() === power`.

File: src/player-data.ts [hunk 4]
```diff
@@ -147,37 +149,68 @@ export default class PlayerData extends EventEmitter {
...
     this.power.area3 += area2ToUp;
     
     //returns real charged power
-    return area1ToUp + area2ToUp;
+    return area1ToUp + area2ToUp + brainstoneUsage;
   }
 
   spendPower(power: number)  {  
+      const totPower = this.power.area3 + this.brainstoneValue() ; 
+      if (this.brainstone === BrainstoneArea.Area3 && ( power >= 3 || ( power<3 && this.power.area3<power))){
```
Comment (coyotte508) [inline src/player-data.ts]: `( power<3 && this.power.area3<power)` can be changed to `this.power.area3 < power`

File: src/enums.ts [hunk 5]
```diff
@@ -278,4 +278,12 @@ export enum FinalTile {
   Gaia= "gaia",
   Sector= "sector",
   Satellite= "satellite"
+}
+
+export enum BrainstoneArea {
+  NotInPlay = "notinplay",
```
Comment (coyotte508) [inline src/enums.ts]: I think `NotInPlay = null` is better

File: src/player-data.ts (hunk 3 above)
Comment (rzulian) [reply src/player-data.ts]: ok

File: src/enums.ts (hunk 5 above)
Comment (rzulian) [reply src/enums.ts]: renamed to Out

**Comment by rzulian** [comment]
This is managing also the ability to select in which order to accept +t and +pw income events. I've managed to have separate events from PI, so I can use activated property to check what's still not collected. I've also modified round boosters and techs to Income event instead of Once, as this is more accurate, even id the round booster are collected once per turn.

File: src/player-data.ts [hunk 6]
```diff
@@ -146,38 +148,69 @@ export default class PlayerData extends EventEmitter {
...
     const area2ToGaia = Math.min(power - area1ToGaia, this.power.area2);
-    const area3ToGaia = Math.min(power - area1ToGaia - area2ToGaia, this.power.area2);
+    const area3ToGaia = Math.min(power - area1ToGaia - area2ToGaia, this.power.area3);
+    const brainstoneNeeded = this.brainstoneInPlay() && this.discardablePowerTokens() === power;
 
     this.power.area1 -= area1ToGaia;
     this.power.area2 -= area2ToGaia;
     this.power.area3 -= area3ToGaia;
+    if (brainstoneNeeded) {
+      this.brainstone = BrainstoneArea.Out;
```
Comment (coyotte508) [inline src/player-data.ts]:
```
this.brainStone = type === Resource.GainTokenGaiaArea ? BrainstoneArea.Gaia : BrainstoneArea.Out
```

Then there's no need for the test below:

```ts
if (brainstoneNeeded) {
   this.brainstone = BrainstoneArea.Gaia;
}
```

File: src/player.ts [hunk 7]
```diff
@@ -207,8 +219,13 @@ export default class Player extends EventEmitter {
...
     if (hex.data.planet !== Planet.Lost) {
-      // Add income of the building to the list of events
-      this.loadEvent(this.board[building].income[this.data[building]]);
+      if (building === Building.PlanetaryInstitute) {
+        // PI has different events
+        this.loadEvents( this.board[Building.PlanetaryInstitute].income );
+      } else {
+        // Add income of the building to the list of events
+        this.loadEvent(this.board[building].income[this.data[building]]);
+        }
```
Comment (coyotte508) [inline src/player.ts]: indent

**Comment by coyotte508** [comment]
I'm sorry, but you got to separate the income pw/t part from the taklons PR.

The taklons PR still doesn't have appropriate victory point calculation, by converting the brainstone to 3c at the end of the game.

And the income pw/t PR has several problems. The activated income events aren't deactivated in the cleanup phase. 

From the rules:

> You choose the order in which you charge and gain power. 
> When you gain income from a source (a single income icon), you 
must **take it all at once** before taking the income from another 
source. 

So there's still a problem. You separated the PI event which gains tokens & power in two, but it shouldn't. It should still be one event, only which reward is gained first should be decided by the player.

The income command should probably be something like:

```ts
// First get income from the two labs (with nevlas, they gain power), then
// the planetary institute (with tokens first), then from the int tech tile,
// then from the eco research track, then from the booster
p1 income lab lab PI(t) tech-int eco faction booster2 lab
```

Actually the income chain only needs to be specified up to when there are still events that gain tokens and events that charge power.

In that last example, the last `lab` can be removed. If it was `booster5` (which charges power) instead of `booster2` (which gains tokens), then the income chain can be simplified further:

```ts
// First get income from the two labs (with nevlas, they gain power), then
// the planetary institute (with tokens first), then from the int tech tile,
// then from the eco research track, then from the booster
p1 income lab lab PI(t)
```

Since all the other events are events that just charge power.

With all that said, it is something that needs careful consideration, maybe even a significant change on the structure / handling of events. It definitely needs **its own PR**.

We also need to consider heuristics. For example if the player has power charged to the max, it's always better to get a power token than to charge power next, and there's no need to ask the player what to do. But it can become complicated with multiple different events, so it can be deferred.

**Comment by rzulian** [comment]
I double checked, and I'm pretty sure that this is the right way to do it. "**single income icon**" means the hand icon, therefore you can select the order in which you want to get the +t and +pw incomes.
The code already manages the use case where you don't have +t and +pw in the incomes, and also in the following phase, if you you don't need to pick incomes (e.g at the beginning you have +4pw, +t, +2pw.  You pick to get +t as first income, everything else is managed by the code).

The only problem I'm seeing is that you get something like
` p1 income 4pw. income t`.

I would prefer to list each income, or at least something like in TM
`# Round 2: income

Chaosmagicians: income
Nomads: income
Mermaids: income. Transform G3 to blue. Transform F2 to blue
Darklings: income
` 

Brainstone for Taklos: https://github.com/donkeytech/gaia-engine/pull/53/files#diff-3239de98447ce2418fd5235e0d690a29R539 should be:
`const resources = pl.data.ores + pl.data.credits + pl.data.qics + pl.data.knowledge + Math.floor((pl.data.power.area2 + (BrainstoneArea.Area2 ? 3 : 0))/ 2) + pl.data.power.area3 + (BrainstoneArea.Area3 ? 3 : 0);`

I can double check the issue you mention on the deactivation.

Let me know what you think about it

**Comment by rzulian** [comment]
I'm sorry, I didn't realised that last commit wasn't in origin yet. Sorry for the confusion

**Comment by coyotte508** [comment]
Regarding https://github.com/donkeytech/gaia-engine/pull/53/commits/64fdfc4abce8e717603aa1a63d79a35d883ce5e2: As mentioned in #20 , under the assumption of #20, it's always beneficial to gain the additional power token for taklons. In the code I see, the player doesn't have a choice anyway, he is always forced to choose to take the token. 

Regarding the order of gaining power & power tokens, it needs to be in its own PR. I won't budge on that. We can discuss the specifics there (or in an issue). I think you're right on the "all at once" question, but even then it still needs to be in its own issue.

**Comment by rzulian** [comment]
Understood. Is there a simple way to move the income phase code into another branch?

**Comment by coyotte508** [comment]
Create another branch from master.

Use `git cherry-pick ffe2da7`. It will add this particular commit to the new branch.

Then, on the **taklons** branch, do `git rebase -i HEAD~5`. Then replace the `pick` for the commit by `drop`. Then `git push --force`.

**Comment by coyotte508** [comment]
There's still the issue about gaining a power token during leech phase.

> Regarding 64fdfc4: As mentioned in the last part of #20 , under the assumption of #20, it's always beneficial to gain the additional power token for taklons. In the current code of the PR, from what I see, the player doesn't have a choice anyway, he is always forced to choose to take the token.

What are your thoughts?

**Comment by rzulian** [comment]
Taklons get +1t only if they leech, and if they leech they get +1t. They can decide the order : leech and get, or get and leech. It can make a difference (eg. you only have the brainstone is in area1, and you leech 1pw. You might want to leech before to take the new token.).

So leech and get are combined. Current code is not managing it.

**Comment by coyotte508** [comment]
So a taklon player, when faced with leeching 2 power, has three choices:

- p1 +t. leech 2
- p1 leech 2 (token is automatically gained after)
- decline 2

I can merge the PR without the freeincome command and this can be added later, or I can wait until everything's ready.

**Comment by rzulian** [comment]
Please wait, working on it. Let's see if the final version works

File: src/player-data.ts [hunk 8]
```diff
@@ -146,38 +148,69 @@ export default class PlayerData extends EventEmitter {
...
     //returns real charged power
-    return area1ToUp + area2ToUp;
+    return area1ToUp + area2ToUp + brainstoneUsage;
   }
 
-  spendPower(power: number)  {  
-      this.power.area3 -= power;
-      this.power.area1 += power;
+  spendPower(power: number) {
+    const totPower = this.power.area3 + this.brainstoneValue();
```
Comment (coyotte508) [inline src/player-data.ts]: this variable is unused

File: src/faction-boards/taklons.ts [hunk 9]
```diff
@@ -1 +1,5 @@
-export default {};
\ No newline at end of file
+import {BrainstoneArea} from '../enums';
+
+export default {
+  brainstone: BrainstoneArea.Area1 
+  };
```
Comment (coyotte508) [inline src/faction-boards/taklons.ts]: indent

File: src/player-data.ts (hunk 6 above)
Comment (rzulian) [reply src/player-data.ts]: ok

File: src/player-data.ts (hunk 8 above)
Comment (rzulian) [reply src/player-data.ts]: ok

File: src/faction-boards/taklons.ts (hunk 9 above)
Comment (rzulian) [reply src/faction-boards/taklons.ts]: ok

**Comment by rzulian** [comment]
Should be fine now. Now the leech command can be Leech xPw,1t or Leech 1t,xPw. This will solve the rewards in the exact order.

File: src/player-data.ts [hunk 10]
```diff
@@ -148,35 +148,40 @@ export default class PlayerData extends EventEmitter {
...
       brainstoneUsage += 1;
       power -= 1;
-      this.brainstone = BrainstoneArea.Area3;
+      brainstonePos = BrainstoneArea.Area3;
     }
 
     const area2ToUp = Math.min(power - area1ToUp, this.power.area2 + area1ToUp);
 
+    if (apply) {
     this.power.area1 -= area1ToUp;
```
Comment (coyotte508) [inline src/player-data.ts]: indent

File: src/engine.spec.ts [hunk 11]
```diff
@@ -593,6 +593,58 @@ describe("Engine", () => {
...
+        p1 build ts -4x-1.
+        p2 leech 1pw
+        p2 build ts -3x-2.
+        p1 leech 2pw
+        p1 build ts -4x2.
+        p2 leech 1pw
+        p2 build PI -3x-2.
+        p1 leech 2pw
+        p1 build lab -4x-1. tech gaia.
+        p2 leech 3pw,1t
```
Comment (coyotte508) [inline src/engine.spec.ts]: The two test can be combined into one.

Remove that last line in the move list. Then:

```ts
expect(() => new Engine([...moves, "p2 leech 3pw,1t"])).to.not.throw();
expect(() => new Engine([...moves, "p2 leech 1t,3pw"])).to.not.throw();
```

File: src/engine.spec.ts [hunk 12]
```diff
@@ -593,6 +593,58 @@ describe("Engine", () => {
       expect(engine.player(Player.Player1).data.advTechTiles).to.include(engine.advTechTiles[AdvTechTilePos.GaiaProject].tile);
       expect(engine.advTechTiles[AdvTechTilePos.GaiaProject].numTiles).to.equal(0);
     });
+
+    it("should allow Taklons to leech with +t freeIncome", () => {
```
Comment (coyotte508) [inline src/engine.spec.ts]: The test is in the "tech tile" portion of the test.

Seee https://travis-ci.org/donkeytech/gaia-engine/builds/402973639?utm_source=github_status&utm_medium=notification : 

```
    tech tiles
      ✓ should prevent picking the same tech tile twice (49ms)
      ✓ should allow to upgrade research area after building a RL, pick tech in nav
      ✓ should allow to get an advanced tech tiles when conditions are met (82ms)
      ✓ should allow Taklons to leech with +t freeIncome (42ms)
      ✓ should allow Taklons to leech with +t freeIncome (53ms)
```

File: src/engine.spec.ts (hunk 11 above)
Comment (rzulian) [reply src/engine.spec.ts]: ok

File: src/engine.spec.ts (hunk 12 above)
Comment (rzulian) [reply src/engine.spec.ts]: ok

**Pull request merged**